File: leetcode/1004.max-consecutive-ones-iii.cpp
/*
 * @lc app=leetcode id=1004 lang=cpp
 *
 * [1004] Max Consecutive Ones III
 */

// @lc code=start

class Solution {
public:
	int longestOnes(vector<int>& nums, int k) {
		int	len = 1;
		int	zeroCount = 0;
		int	maxSubarray = 0;
		int	numsLen = nums.size();
		int	i = 0;
		int j = 0;

		while (j < numsLen) {
			zeroCount += (nums[j] == 0);
			while (zeroCount > k) {
				if (nums[i] == 0)
					zeroCount--;
				i++;
			}
			len = j - i + 1;
			maxSubarray = len > maxSubarray ? len : maxSubarray;
			j++;
		}
		return maxSubarray;
	}
};
// @lc code=end

File: leetcode/1004.max-consecutive-ones-iii.py
#
# @lc app=leetcode id=1004 lang=python3
#
# [1004] Max Consecutive Ones III
#

# @lc code=start
class Solution:
	def longestOnes(self, nums: List[int], k: int) -> int:
		i, maxSubarray, zeroCount = 0, 0, 0
		for j, val in enumerate(nums):
			zeroCount += (val == 0)
			while zeroCount > k:
				zeroCount -= (nums[i] == 0)
				i += 1
			maxSubarray = max(maxSubarray, j - i + 1)
		return maxSubarray
# @lc code=end

